File: UnionFindSet/src/main/java/Solution684.java
public class Solution684 {
    int[] fa;

    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        fa = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            fa[i] = i;
        }

        for (int[] edge : edges) {
            if (find(edge[0]) == find(edge[1])) {
                return edge;
            }

            unionSet(edge[0], edge[1]);
        }
        return edges[0];
    }

    private int find(int x) {
        if (x == fa[x]) {
            return x;
        }
        return fa[x] = find(fa[x]);
    }

    private void unionSet(int x, int y) {
        x = find(x);
        y = find(y);

        if (x != y) {
            fa[x] = y;
        }
    }
}

File: UnionFindSet/src/main/python/solution684.py
from typing import List


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        m = len(edges)
        self.fa = [i for i in range(m + 1)]

        for edge in edges:
            if self.find(edge[0]) == self.find(edge[1]):
                return edge

            else:
                self.union_set(edge[0], edge[1])

    def find(self, x):
        if x == self.fa[x]:
            return x

        self.fa[x] = self.find(self.fa[x])

        return self.fa[x]

    def union_set(self, x, y):
        x = self.find(x)
        y = self.find(y)

        if x != y:
            self.fa[x] = y
